### ISSUE TITLE
Static inside for breaks models

### ISSUE THREAD

**Issue by wsnyder** [verilator/verilator#4027]
@RRozak adding #4018 is causing the verilator_ext_tests and specifically SweRV to fail.  It has this:

 ```
              for (int i=0; i<2; i++)
                    if (trace_rv_i_valid_ip[t][i]==1) begin
                        bit i0;
                        i0 = i != 0 || trace_rv_i_valid_ip[t]!=3 && wb_tid[1] == t && wb_valid[1];
```

If this fails likely other users will also be impacted.  Do you think this can get either supported (ideally), backed out, or I suppose look to see if the variable is set in the immediate code path below unconditionally?

**Comment by wsnyder** [comment]
I found some other models that also do this, it seems relatively common. In each case I think "automatic" was the intent.

Another idea: Add a new warning that this is being converted to automatic, thus most code should work.  Add a note to the description that in future versions Verilator may correctly treat this as static, but because this is a common mistake the warning will still exist.

**Comment by RRozak** [comment]
Thanks for the information. I will look at it.

**Comment by RRozak** [comment]
The approach I tried is to wrap a declaration assignment into AstInitialStatic and move it to the top of a module.

It is more complicated in a `do..while` loop, because the body is copied and put before a loop. In such a case I wanted to move declaration assignment to the top of a module and wrap it into AstInitialStatic (like in any other loop). Then remove a declaration assignment from the block that was put before a loop and replace references with the ones that point to static variables in a loop.

There were some problems with that approach. References are linked later and we need to have it linked to move AstInitialStatic. We can create hierarchical references instead, but names to unnamed blocks are assigned later, in V3LinkDot too. In addition, the handling of unnamed blocks doesn't work correctly, which I want to fix (I am working on it currently). So far, I prepared the conversion to automatic: https://github.com/verilator/verilator/pull/4030.

Another approach is to add a field to an assignment that would mean that it is a declaration assignment to a static variable and keep that assignment in a loop until it can be wrapped into AstInitialStatic and moved under a module. Maybe it is better and easier way.
